### PR TITLE
Product list routes

### PR DESCRIPTION
**Pull request by davidch25**
### What this PR does (required):
- added routes & controllers for product and list collections
- controllers include adding, removing, updating, and getting documents

### Screenshots / Videos (required):
- Adding a list
https://gyazo.com/ac55300d08a9c82812489de2d99be4b1
- Updating a list
https://gyazo.com/a21b6c925b82fbbdcbeeb59faf913444
- Getting a list
https://gyazo.com/0a4bd9c47135e46bd087063ca021fd06
- Removing a list
https://gyazo.com/611b36e16d91f3a3fd50ce2cc3cce3b3

### Any information needed to test this feature (required):
The routes for products are:
- product/addProduct
- product/removeProduct/:id
- product/updateProduct/:id
- product/findProducts/:id

The routes for lists are:
- list/addList
- list/removeList/:id
- list/updateList/:id
- list/findLists/:userId

note that findList uses userId, and not the object's id because we need to find all lists belonging to the specified userId

### CODE REVIEW

**Comment by davidch25** [comment]
this branch actually has the changes from product-list-models branch because I needed those files for testing purposes

File: server/app.js [hunk 1]
```diff
@@ -1,71 +1,75 @@
...
-const path = require("path");
-const http = require("http");
-const express = require("express");
-const socketio = require("socket.io");
-const { notFound, errorHandler } = require("./middleware/error");
-const connectDB = require("./db");
-const { join } = require("path");
-const cookieParser = require("cookie-parser");
-const logger = require("morgan");
+const colors = require('colors')
```
Comment (dproc96) [inline server/app.js]: I think you need to align your formatter settings, check with your team about this

File: server/controllers/product.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+const Product = require('../models/Product')
+const asyncHandler = require('express-async-handler')
+
+//add a list to the collection
+//Route POST /product/addProduct
+exports.addProduct = asyncHandler(async (req, res, next) => {
+  Product.create(req.body).then(function (product) {
+    res.send(product)
+  })
```
Comment (dproc96) [inline server/controllers/product.js]: You should also be adding a product to a list here, right?

File: server/controllers/product.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
+const Product = require('../models/Product')
+const asyncHandler = require('express-async-handler')
+
+//add a list to the collection
```
Comment (dproc96) [inline server/controllers/product.js]:
```suggestion
//add a product to the collection
```

File: server/controllers/product.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    product.save()
+    res.send({ data: product })
+  } catch {
+    res.status(404).send({ error: 'Product not found' })
+  }
+})
+
+//find the product by id parameter
+//Route GET /product/findProducts/:id
+exports.findProducts = asyncHandler(async (req, res, next) => {
```
Comment (dproc96) [inline server/controllers/product.js]:
```suggestion
exports.findProduct = asyncHandler(async (req, res, next) => {
```
If the function name is plural, the expected return would be an array of values

File: server/controllers/product.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  try {
+    const products = await Product.findById(req.params.id)
```
Comment (dproc96) [inline server/controllers/product.js]: Ditto here with the variable name

File: server/controllers/list.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
+const List = require('../models/List')
+const asyncHandler = require('express-async-handler')
+
+//add a list to the collection
+//Route POST /list/addList
+exports.addList = asyncHandler(async (req, res, next) => {
+  List.create(req.body).then(function (list) {
+    res.send(list)
+  })
```
Comment (dproc96) [inline server/controllers/list.js]: There should be a linkage established here with the logged in user

File: server/controllers/list.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  } catch {
+    res.status(404).send({ error: 'List not found' })
+  }
+})
+
+//find the lists by userId parameter
+//Route GET /list/findLists/:userId
+exports.findLists = asyncHandler(async (req, res, next) => {
+  try {
+    const lists = await List.find({ userId: req.params.userId })
```
Comment (dproc96) [inline server/controllers/list.js]: Is there a use case for fetching someone else's lists? If not, you could just use the req.user to remove the possibility someone could access another user's list

File: server/routes/list.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
+const express = require('express')
+const router = express.Router()
+const {
+  addList,
+  removeList,
+  updateList,
+  findLists,
+} = require('../controllers/list')
+
+router.route('/addList').post(addList)
```
Comment (dproc96) [inline server/routes/list.js]: All of these routes should probably be protected

**Review by dproc96 — COMMENTED** [review]
Had a few notes about implementation, and I also think your formatter setting might be misaligned with the rest of the group. 

Also if you want the changes from the model branch to not show in the diff, you can make the base branch for this PR the model branch (go to edit and select from the base dropdown)

File: server/controllers/list.js (hunk 6 above)
Comment (davidch25) [reply server/controllers/list.js]: By linkage do you mean that each user object should have an array of lists and when I add a list that array should be updated on the user object? Currently my models are such that lists have the ObjectId of the user which is required.

Comment (dproc96) [reply server/controllers/list.js]: Neither of those are happening here though, it's just creating a list. Unless you're passing the userId as part of the request body, which is a security concern as any user could create a list on someone else's account

File: server/routes/list.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  removeList,
+  updateList,
+  findList
+} = require("../controllers/list");
+
+router.route("/addList").post(protect, addList);
+
+router.route("/removeList/:id").delete(protect, removeList);
+
+router.route("/updateLis/:id").patch(protect, updateList);
```
Comment (bonnieli) [inline server/routes/list.js]:
```suggestion
router.route("/updateList/:id").patch(protect, updateList);
```

**Review by bonnieli — APPROVED** [review]

**Pull request merged**